### ISSUE TITLE
Error when using brewcask

### ISSUE THREAD

**Issue by jordigg** [TheKevJames/puppet-homebrew#2]
I'm using this module to get Brew and Brewcask on MacOS. Brew works fine but I can't make brewcask to work. I'm using Puppet version 4.

I'm using the package like in the examples, my code is below.
```
$brewcask_packages = ['java', 'sbt', 'cmake']
package { $brewcask_packages:
	ensure		=> present,
	provider	=> 'brewcask',
}
```
I get always the same error regardless which package I'm trying to install.
```
Debug: Executing: '/usr/bin/stat -nf %Uu /usr/local/bin/brew'
Debug: Executing: '/usr/bin/stat -nf %Ug /usr/local/bin/brew'
Debug: Executing with uid=0 gid=80: '/usr/local/bin/brew cask list --versions'
**Error: /Package[cmake]: Could not evaluate: no implicit conversion of nil into String**
```

When `brew cask list --versions cmake` is run manually returns the following: `Warning: cmake is not installed` `Error: nothing to list` but `brew list --versions python` returns `Error: No such keg: /usr/local/Cellar/python` when the packages aren't installed.

Any clues?

Thanks!

**Comment by nexussafe** [comment]
I have the same issue. I am trying to install 'chrome', 'firefox', and 'thunderbird' packages

**Comment by jordigg** [comment]
Well the error is that we try to parse an empty result when nothing is present and the following code fails.

```
result = Hash[result.lines.map {|line| line.split}]
result = name + ' ' + result[name]
```

to prevent that I have added a condition to only do it when there's content we can parse. Like so:

```
if !result.empty?
    result = Hash[result.lines.map {|line| line.split}]
    result = name + ' ' + result[name]
end
```

This way works. I'll check further on once the package is already installed and see if the results are being parsed OK. Check the code [here](https://github.com/improbable-io/puppet-homebrew/commit/0feb76390b1b05ff586c630dc47a3c3a70f8cded).

@TheKevJames Please take a look if works for you too.

**Comment by TheKevJames** [comment]
@nexussafe @jordigg I've cut a new release with my fixes (release version 1.2.0). Thanks again for helping to track down this issue. I'm marking this issue as closed -- if anything else comes up, please open a new issue.

**Comment by nexussafe** [comment]
Unfortunately it still does not work for me. It fails in the other place:

    Debug: Prefetching brewcask resources for package
    Debug: Executing '/usr/bin/stat -nf %Uu /usr/local/bin/brew'
    Debug: Executing '/usr/bin/stat -nf %Ug /usr/local/bin/brew'
    Debug: Executing '/usr/local/bin/brew cask list --versions'
    Debug: Executing '/usr/bin/stat -nf %Uu /usr/local/bin/brew'
    Debug: Executing '/usr/bin/stat -nf %Ug /usr/local/bin/brew'
    Debug: Executing '/usr/local/bin/brew cask list --versions'
    Debug: Executing '/usr/bin/stat -nf %Uu /usr/local/bin/brew'
    Debug: Executing '/usr/bin/stat -nf %Ug /usr/local/bin/brew'
    Debug: Executing '/usr/local/bin/brew cask install firefox'
    Error: Could not find package firefox
    Error: /Stage[main]/Main/Package[firefox]/ensure: change from absent to present failed: Could not find package firefox`

**Comment by nexussafe** [comment]
I found a workaround for my problem, `brewcommon.rb:17`

      def self.execute(cmd)
        owner = stat('-nf', '%Uu', '/usr/local/bin/brew').to_i
        group = stat('-nf', '%Ug', '/usr/local/bin/brew').to_i
        home  = Etc.getpwuid(owner).dir

        super(cmd, :uid => owner, :gid => group,
              :custom_environment => { 'HOME' => home })
      end

I deleted this 2 arguments `:uid => owner, :gid => group,` and it started working.

            super(cmd, # No idea why
                  :custom_environment => { 'HOME' => home })

**Comment by TheKevJames** [comment]
@nexussafe Interesting, sounds like you may have installed brew as root? Note that [brew does not support this workflow](https://github.com/Homebrew/homebrew/issues/9953). The following info could help us debug:

    $ ls -la /usr/local/bin | grep brew
    -rwxr-xr-x   1 kevin  admin   656B Feb 25 10:57 brew
    $ stat -nf %Uu /usr/local/bin/brew
    501
    $ stat -nf %Ug /usr/local/bin/brew
    80
    $ id
    uid=501(kevin) gid=20(staff) groups=20(staff),<SNIP>,80(admin),<SNIP>
    $ sudo /opt/puppetlabs/bin/puppet apply -e "notify { generate('/usr/bin/stat', '-nf', '%Uu', '/usr/local/bin/brew'): loglevel => err }"
    Notice: Compiled catalog for aerobox.local in environment production in 0.02 seconds
    Error: 501
    Error: /Stage[main]/Main/Notify[501]/message: defined 'message' as '501'
    Notice: Applied catalog in 0.04 seconds
    $ sudo /opt/puppetlabs/bin/puppet apply -e "notify { generate('/usr/bin/stat', '-nf', '%Ug', '/usr/local/bin/brew'): loglevel => err }"
    Notice: Compiled catalog for aerobox.local in environment production in 0.04 seconds
    Error: 80
    Error: /Stage[main]/Main/Notify[80]/message: defined 'message' as '80'
    Notice: Applied catalog in 0.03 seconds

**Comment by nexussafe** [comment]
No, I installed brew like this

    ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"

I also installed puppet via homebrew

    $ ls -la /usr/local/bin/brew
    -rwxr-xr-x  1 nexussafe  admin  656 Mar  8 17:46 /usr/local/bin/brew
    % stat -nf %Uu /usr/local/bin/brew
    501%
    % stat -nf %Ug /usr/local/bin/brew
    80%
    % id
    uid=501(nexussafe) gid=20(staff) groups=20(staff),401(com.apple.sharepoint.group.1),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),33(_appstore),100(_lpoperator),204(_developer),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh)
    % which puppet
    /usr/bin/puppet
    % puppet --version
    3.8.6
    % sudo puppet apply -e "notify { generate('/usr/bin/stat', '-nf', '%Uu', '/usr/local/bin/brew'): loglevel => err }"
    Error: Could not parse for environment production: Syntax error at ':'; expected '}' at line 1 on node nexussafe-mbp.lan
    Error: Could not parse for environment production: Syntax error at ':'; expected '}' at line 1 on node nexussafe-mbp.lan

**Comment by jordigg** [comment]
Kevin's changes work for us. I would recommend installing Puppet first and then everything else.

I have scripts for bootstraping machines for MacOSX, Linus and Windows on my gist. Here for Mac https://gist.github.com/jordigg/be23c98e8adbf37f2d8c

I'm sure is a permission issue with they way you have it all installed. You can install brew and brewcask with this same module.

**Comment by TheKevJames** [comment]
Ah, I understand; since you've installed puppet through homebrew, there are going to be some differences in how it works compared to [how I installed it](http://docs.puppetlabs.com/puppet/latest/reference/install_osx.html). As an aside, you're also a major version of puppet behind. 

I'd recommend upgrading, but I'll also look into what would need to be done to support puppet 3 installed through brew. Thanks for bringing this to my attention! Separated this issue to #4.

**Comment by tfnico** [comment]
Sorry for commenting in a closed issue, but I'm seeing the same problem, even after I installed puppet properly:

```
Debug: Executing: '/usr/bin/stat -nf %Uu /usr/local/bin/brew'
Debug: Executing: '/usr/bin/stat -nf %Ug /usr/local/bin/brew'
Debug: Executing with uid=501 gid=80: '/usr/local/bin/brew cask list --versions'
Debug: Executing: '/usr/bin/stat -nf %Uu /usr/local/bin/brew'
Debug: Executing: '/usr/bin/stat -nf %Ug /usr/local/bin/brew'
Debug: Executing with uid=501 gid=80: '/usr/local/bin/brew cask install vlc'
Error: Could not find package vlc
Error: /Stage[main]/Tfnico::Packages/Package[vlc]/ensure: change from absent to present failed: Could not find package vlc
.....

Thomass-MBP-2:prefs tfnico$ ls -la /usr/local/bin | grep brew
-rwxr-xr-x   1 tfnico  admin  656 Mar 14 16:35 brew

Thomass-MBP-2:prefs tfnico$ stat -nf %Uu /usr/local/bin/brew
501

Thomass-MBP-2:prefs tfnico$ stat -nf %Ug /usr/local/bin/brew
80

Thomass-MBP-2:prefs tfnico$ id
uid=501(tfnico) gid=20(staff) groups=20(staff),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),701(com.apple.sharepoint.group.1),33(_appstore),100(_lpoperator),204(_developer),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh)

Thomass-MBP-2:prefs tfnico$ sudo /opt/puppetlabs/bin/puppet apply -e "notify { generate('/usr/bin/stat', '-nf', '%Uu', '/usr/local/bin/brew'): loglevel => err }"
Password:
Notice: Compiled catalog for thomass-mbp-2.lan in environment production in 0.05 seconds
Error: 501
Error: /Stage[main]/Main/Notify[501]/message: defined 'message' as '501'
Notice: Applied catalog in 0.01 seconds

Thomass-MBP-2:prefs tfnico$ sudo /opt/puppetlabs/bin/puppet apply -e "notify { generate('/usr/bin/stat', '-nf', '%Ug', '/usr/local/bin/brew'): loglevel => err }"
Notice: Compiled catalog for thomass-mbp-2.lan in environment production in 0.06 seconds
Error: 80
Error: /Stage[main]/Main/Notify[80]/message: defined 'message' as '80'
Notice: Applied catalog in 0.02 seconds

Thomass-MBP-2:prefs tfnico$ puppet module list
/Users/tfnico/.puppetlabs/etc/code/modules
├── puppetlabs-stdlib (v4.11.0)
└── thekevjames-homebrew (v1.2.0)
/opt/puppetlabs/puppet/modules (no modules installed)

Thomass-MBP-2:prefs tfnico$ puppet --version
4.3.2
```

**Comment by TheKevJames** [comment]
Hi! No worries about commenting on a closed issue, it looks like it might be related. It looks like you may have installed this module without privileges (since the module looks to have been installed into your home directory).

Could you please run

    /opt/puppetlabs/bin/puppet module uninstall thekevjames-homebrew
    sudo /opt/puppetlabs/bin/puppet module install thekevjames-homebrew

And then retry your module? (Also, if you could paste the command your using / module you're running, that'd be great).

**Comment by tfnico** [comment]
Thanks for the quick reply. It wasn't clear to me which commands should be run privileged, and which not. I first tried this:

```
➜  ~ /opt/puppetlabs/bin/puppet module uninstall --ignore-changes thekevjames-homebrew
Notice: Preparing to uninstall 'thekevjames-homebrew' ...
Removed 'thekevjames-homebrew' (v1.2.0) from /Users/tfnico/.puppetlabs/etc/code/modules
➜  ~ sudo /opt/puppetlabs/bin/puppet module install thekevjames-homebrew
Password:
Notice: Preparing to install into /etc/puppetlabs/code/environments/production/modules ...
Notice: Downloading from https://forgeapi.puppetlabs.com ...
Notice: Installing -- do not interrupt ...
/etc/puppetlabs/code/environments/production/modules
└── thekevjames-homebrew (v1.2.0)
➜  ~ cd pr
cd: no such file or directory: pr
➜  ~ cd prefs
➜  prefs git:(master) ✗ ls
README.md           deploy.sh           env                 global-git-ignores  puppet-mac.pp       tmux.conf           vimrc.after
ackrc               diffs.sh            git-completion.bash profile             reverse-deploy.sh   vim-setup.sh        vimrc.before
bin-source          dir_colors          gitconfig           puppet-linux.pp     subversion-config   vimrc               zshrc
➜  prefs git:(master) ✗ puppet apply puppet-mac.pp
zsh: command not found: puppet
➜  prefs git:(master) ✗ /opt/puppetlabs/puppet/bin/puppet apply puppet-mac.pp
Notice: Compiled catalog for thomass-mbp-2.lan in environment production in 0.06 seconds
Error: Parameter provider failed on Package[curl]: Invalid package provider 'brew' at /Users/tfnico/prefs/puppet-mac.pp:23
```

**Comment by nexussafe** [comment]
Try

    sudo /opt/puppetlabs/puppet/bin/puppet apply puppet-mac.pp

**Comment by tfnico** [comment]
@nexussafe Thanks, I just updated my reply, which got submitted prematurely :)

**Comment by TheKevJames** [comment]
That's the gist of it -- in general, always executing puppet commands as root (ie. with sudo) is a good idea.

**Comment by tfnico** [comment]
Thanks for confirming that. Here, a blog-post in your honor: http://blog.tfnico.com/2016/03/replacing-boxen-with-vanilla-puppet-for.html :)

**Comment by steveames** [comment]
Hi. Checking for empty seems insufficien.t In my case I'm not getting an empty but a 'Warning: nothing to list'. This change seems to get me where I'm going WRT the original issue. Possibly a check for empty may also be required but it was not in my case. OSX 10.11, Puppet 4, brew installed via this module. With this change puppet didn't error out and my cask modules got installed.

brewcask.rb starting around line 76

    begin
      if name = options[:justme]
        # Of course brew-cask has a different --versions format than brew when
        # getting the version of a single package
        tresult = execute([command(:brew), :cask, :list, '--versions'])

        if tresult.split.first == 'Warning:'
          result = ''
        else
          tresult = Hash[tresult.lines.map {|line| line.split}]
          if tresult[name] != nil
            result = name + ' ' + tresult[name]
          else
            result = ''
          end
        end
      else
        result = execute([command(:brew), :cask, :list, '--versions'])
      end

**Comment by TheKevJames** [comment]
My bad, didn't see the above comment. Reopening so I remember to look into this.

**Comment by TheKevJames** [comment]
This should be fixed by #51 given we no longer parse output.
